Refactored archiving

DavidHuber-NOAA opened the pull request:
# Description
This provides a new pygfs task, archive.py, that provides all of the tools necessary to archive data to the local (`ARCDIR`) and backup (`ATARDIR`) archive directories.  YAML-Jinja2 templates are provided to define the file to be archived or tarred to replace the `hpssarch_gen.sh`, `exglobal_earc`, and `exglobal_archive.sh` scripts and make it easier to add new data and explicitly handle optional and required files.

For `ATARDIR` archiving, a master jinja template is provided for each `RUN` (i.e. master_gdas.yaml.j2, master_gfs.yaml.j2, master_enkf.yaml.j2).  The master_enkf.yaml.j2 template is used for both `enkfgdas` and `enkfgfs` `RUN`s.  These templates then include the appropriate `RUN`-specific jinja templates (e.g. gdas.yaml.j2) based on experiment, cycle, and coupled parameters.  Each of these templates corresponds to a single tarball to populate and are tabbed 4 spaces so they are defined within the master `datasets` dictionary.

Future developers should not have to make modifications to archive.py unless archiving is being enabled for a new `RUN` (e.g. `gefs`) and then only a single `elif` needs to be added to the configure method to specify the master `ATARDIR` template to archive (e.g. `master_gefs.yaml.j2`).  If a new component is coming online that needs to be archived to `ATARDIR` (e.g. SNOW), then create a new template for each `RUN` that it needs to be archived for (e.g. `gdassnow.yaml.j2`) and reference the template in the appropriate master templates, e.g.

`master_gdas.yaml:`
```jinja
{% if DO_SNOW %}
{% include "gdassnow.yaml.j2" %}
{% endif %}
```

A few other issues were addressed along the way:
1. Aerosols have been reenabled.  Aerosol forecasts should only be performed during gdas cycles, but analyses can be performed for both gfs and gdas cycles.  This was accomplished by setting separate `AERO_<job>_CDUMP` variables to parse on for both `ANL` and `FCST` jobs.
2. Fixed the name of the `cice6_rst_ok` variable in `forecast_det.sh`.  This prevented restarts from being used for cice-enable experiments.  This feature was not tested.
3. Create a temporary fix for the `wgrib` utility.  For spack-stack 1.6.0, the `grib-util` module file does not declare `WGRIB`.  An issue is open (https://github.com/JCSDA/spack-stack/issues/1097) to fix this in existing installations.  Once complete, this temporary fix should be removed.
4. The number of `earc` jobs has been reduced for lower resolution experiments.  Both C48 and C96 experiments will now only have two earc jobs (one for the non-member files to archive and another for the member files).  C192 will have up to 3 earc jobs (one non-member, one for members 1-40 and another for members 41-80, if needed).

Resolves #2345 #2318 

# Type of change
- Bug fix (fixes something broken)
- New feature (adds functionality)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? NO

# How has this been tested?
- [x] GDAS ATM cycling on Hera
- [x] GFS ATM cycling on Hera
- [x] ENKFGDAS on Hera
- [x] GFS forecast-only S2SW
- [x] GDAS cycling with JEDI
- ~GFS cycling with JEDI~ Not currently possible, will test later.
- [x] GEFS forecast - GEFS does not perform archiving yet, but the forecast did run successfully.
- [x] Local ATM forecast-only

# Checklist
- [x] Any dependent changes have been merged and published
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes

DavidHuber-NOAA committed:
Initial set of archiving yamls.

DavidHuber-NOAA committed:
rework gdas.yaml to Jinja template

DavidHuber-NOAA committed:
Restart with a blank slate; rename gdas.yaml.

DavidHuber-NOAA committed:
Create archive task class with initial gdas.yaml.j2

This also
- adds a missing COM variable to JGLOBAL_ARCHIVE
- Creates a test driver script and python program

DavidHuber-NOAA committed:
Create configure method

DavidHuber-NOAA committed:
Updates to archive task.

DavidHuber-NOAA committed:
Add htar features to archive task, update gdas template. (untested)

DavidHuber-NOAA committed:
Merge remote-tracking branch 'emc/develop' into feature/archive_yaml

DavidHuber-NOAA committed:
Merge remote-tracking branch 'emc/develop' into feature/archive_yaml

DavidHuber-NOAA committed:
Update archive protocol function, add templates.

DavidHuber-NOAA committed:
Merge remote-tracking branch 'emc/develop' into feature/archive_yaml

DavidHuber-NOAA committed:
Archive protocol overhaul.

- Use new hsi/htar classes in wxflow
- Assign cycle dates/hours in the job to avoid assigning in each template
- Rename templates to match current/EE2 naming conventions
- Write function to relative-ize COM paths

DavidHuber-NOAA committed:
Merge remote-tracking branch 'emc/develop' into feature/archive_yaml

DavidHuber-NOAA committed:
Update snowda archiving to match develop.

DavidHuber-NOAA committed:
Address PEP8 issues.

DavidHuber-NOAA committed:
Add gdasocean_analysis template

DavidHuber-NOAA committed:
Add local (ARCDIR) archiving to archive.py.

DavidHuber-NOAA committed:
Fix PEP8 errors.

DavidHuber-NOAA committed:
Let archive.py handle tracker pslot renaming.

DavidHuber-NOAA committed:
Rewrite archive job to use exglobal_archive.py.

DavidHuber-NOAA commented:
This draft only contains local archiving to the `ARCDIR` and GDAS backup archiving to `ATARDIR`, but this will allow review of the work so far while continuing to GFS and ENKF archiving.

DavidHuber-NOAA committed:
Enabled GFS archiving and reworked some features

- Moved the cycle_YYYYMMDDHH, etc, variable declarations to templates
- Use COM directories inside of the templates (instead of _dir directories)
- Add for-looping for tile numbers
- Added several new templates needed for the GFS
- Improved ATARDIR archiving logic

DavidHuber-NOAA committed:
Update wxflow hash.

DavidHuber-NOAA committed:
Add enkf archiving.

DavidHuber-NOAA committed:
Remove now-defunct hpssarch_gen.sh.

DavidHuber-NOAA committed:
Address PEP8 warnings

DavidHuber-NOAA committed:
Fix additional PEP8 warnings

DavidHuber-NOAA committed:
Merge remote-tracking branch 'emc/develop' into feature/archive_yaml

DavidHuber-NOAA committed:
Fixes for snow DA archiving.

DavidHuber-NOAA committed:
Remove unused fhmin.

DavidHuber-NOAA committed:
Better explicitize enkf archiving

DavidHuber-NOAA committed:
Further explicitize archiving templates.

DavidHuber-NOAA committed:
Small fix for snow DA archiving.

DavidHuber-NOAA committed:
Fix TC genesis arcdir archiving.

DavidHuber-NOAA committed:
Merge remote-tracking branch 'emc/develop' into feature/archive_yaml

DavidHuber-NOAA committed:
Fix pgrb2b analysis filename

DavidHuber-NOAA committed:
Fix cyclone tracker filenames.

DavidHuber-NOAA committed:
Move forecast hours and restart intervals to config.base.

DavidHuber-NOAA committed:
Archive all sfcf* files.

DavidHuber-NOAA committed:
Refactor restart/forecast hour-based archiving

DavidHuber-NOAA committed:
Add TODOs for GDAS component archiving

DavidHuber-NOAA committed:
Explicitly archive gfs_collective files

DavidHuber-NOAA committed:
More explicitly archive gridded GFS Wave products

DavidHuber-NOAA committed:
Add optional wave station products to archives

DavidHuber-NOAA committed:
Fix restart intervals in archiving templates.

DavidHuber-NOAA committed:
Address shellcheck suggestion

DavidHuber-NOAA committed:
Fix enkf_grp optional file declarations.

DavidHuber-NOAA committed:
Move verfrad warning files to optional list

DavidHuber-NOAA committed:
Fix restart_interval loop control.

DavidHuber-NOAA committed:
Copy relevant changes to gefs config files.

DavidHuber-NOAA committed:
Remove amsua files from archive list

DavidHuber-NOAA committed:
Merge remote-tracking branch 'emc/develop' into feature/archive_yaml

DavidHuber-NOAA committed:
Fix wave FHMAX for GEFS.

DavidHuber-NOAA committed:
Added untested enkfgfs warning to archive.py

DavidHuber-NOAA committed:
Merge branch 'feature/archive_yaml' of github.com:DavidHuber-NOAA/global-workflow into feature/archive_yaml

DavidHuber-NOAA committed:
Remove debug print statement

DavidHuber-NOAA committed:
Remove 'untested' warning

DavidHuber-NOAA committed:
Corrected genesis COM path

DavidHuber-NOAA committed:
Remove unsaved atm_stoch files

DavidHuber-NOAA committed:
Move restart_interval* to more logical locations

DavidHuber-NOAA commented:
This PR has been thoroughly tested on Hera and is now ready for review.

Some TODOs remain and issues will be opened to address these:
- GDAS-based analyses could not be tested at the time of this PR.  As it is tested, the appropriate templates should be fleshed out as many of them are grabbing whole COM directories. #2539 
- Some log files are archived using wildcards.  This should be reworked so only the expected log files are archived. #2539 
- Local tarring (i.e. when `LOCALARCH=YES`) uses tar functions within the Archive class.  These should be moved to wxflow and generalized.

DavidHuber-NOAA committed:
Remove unused gfs_restartb template

DavidHuber-NOAA commented:
Following a conversation with @aerorahul, I'm converting this back to draft while I move much of the logic in the archive.py configure method into the jinja yamls.

DavidHuber-NOAA committed:
Add master template outline.

DavidHuber-NOAA committed:
Adjust earc groups by resolution

DavidHuber-NOAA committed:
Fix template constants in jinja

DavidHuber-NOAA committed:
Rename save_restart variables.

DavidHuber-NOAA committed:
Fix atardir datasets

DavidHuber-NOAA committed:
Merge remote-tracking branch 'emc/develop' into feature/archive_yaml

DavidHuber-NOAA committed:
Update wxflow/jinja

DavidHuber-NOAA committed:
Fix iaufhrs typo

DavidHuber-NOAA committed:
Fix timedelta typo

DavidHuber-NOAA committed:
Apply fixes to master templates

DavidHuber-NOAA committed:
Fix dataset parsing

DavidHuber-NOAA committed:
Add FATAL ERROR to fatal error messages

DavidHuber-NOAA committed:
Fix whitespace

CatherineThomas-NOAA commented on parm/archive/master_gdas.yaml.j2:
```diff
@@ -0,0 +1,39 @@
+{% include "gdasocean_analysis.yaml.j2" %}
+{% endif %}
+
+{% if DO_ICE %} {% include "gdaswave.yaml.j2" %} {% endif %}
```
Should this be DO_WAVE instead of DO_ICE?

DavidHuber-NOAA replied:
@CatherineThomas-NOAA Yes, it should.  Thanks!

DavidHuber-NOAA replied:
Fixed

DavidHuber-NOAA committed:
Update gdas component head names

DavidHuber-NOAA committed:
Flat-layer the master templates, fix wave archiving

DavidHuber-NOAA committed:
Merge remote-tracking branch 'emc/develop' into feature/archive_yaml

DavidHuber-NOAA committed:
Correct CICE namelist.

DavidHuber-NOAA committed:
Add quick fix for wgrib.

DavidHuber-NOAA committed:
Replace bash templates in Jinja templates.

DavidHuber-NOAA committed:
First attempt at reenabling aerosols

DavidHuber-NOAA committed:
Fix snow diag tarring

DavidHuber-NOAA committed:
Fix build_all comment

DavidHuber-NOAA committed:
Fix cice restart variable name

DavidHuber-NOAA committed:
Merge remote-tracking branch 'emc/develop' into feature/archive_yaml

DavidHuber-NOAA committed:
Transfer gfs config fixes to gefs configs

DavidHuber-NOAA committed:
Fix gdas_tasks

DavidHuber-NOAA commented:
Testing has been completed on all CI tests and this PR is now ready for review.  I will be updating the description to include a few other minor and major fixes going in with this.  The gist:
- reenable aerosol analyses
- fix the name of a cice6 restart variable
- a temporary fix for wgrib
- reduce the number of earc groups (more members/job) at lower res

CoryMartin-NOAA commented on parm/config/gfs/config.fcst:
```diff
@@ -13,6 +13,12 @@ case ${WAVE_CDUMP} in
   *) DO_WAVE="NO" ;; # Turn waves off
 esac
 
+# Turn off waves if not used for this CDUMP
```
aero not waves

DavidHuber-NOAA replied:
Thanks, fixed.

CoryMartin-NOAA commented on parm/config/gfs/config.base:
```diff
@@ -227,6 +230,8 @@ case "${APP}" in
     ;;
   ATMA)
     export DO_AERO="YES"
+    export AERO_ANL_CDUMP="gdas"
```
I think we want the default ATMA behavior to be `AERO_ANL_CDUMP="both"` and `FCST="gdas"` if I understand these correctly. Basically, we want the analysis in both GFS and GDAS but only the GDAS fcst has aerosols in it.

DavidHuber-NOAA replied:
OK, I've turned it on for both and will test for the gfs cycle.

CoryMartin-NOAA replied:
I don't remember if we have "gfs" obs for the test or not, so if it fails, we can revert it back.

DavidHuber-NOAA replied:
It works!

DavidHuber-NOAA committed:
Fix aerosol comment

DavidHuber-NOAA committed:
Turn aerosol analyses on for gfs and gdas runs

CatherineThomas-NOAA commented on parm/archive/master_gdas.yaml.j2:
```diff
@@ -0,0 +1,55 @@
+{% include "gdasice.yaml.j2" %}
+{% endif %}
+
+{% if DO_OCEAN %}
```
Should this be DO_OCN?

DavidHuber-NOAA replied:
Yes it should.  I'll need to rerun the `AOWCDA` test case to verify.

DavidHuber-NOAA replied:
Fixed.

CatherineThomas-NOAA commented on parm/archive/master_gdas.yaml.j2:
```diff
@@ -0,0 +1,55 @@
+
+{% if DO_OCEAN %}
+{% include "gdasocean.yaml.j2" %}
+{% include "gdasocean_analysis.yaml.j2" %}
```
@DavidHuber-NOAA We have an update for the gdasocean_analysis tarball in PR #2560 ([commit](https://github.com/NOAA-EMC/global-workflow/pull/2560/commits/9588c9c5e646b77e676a6e2d08c37b6d9d3dd17a)).  We changed the trigger from DO_OCN to DO_JEDIOCNVAR as we have run the case where the forecast uses the ocean model but there is no ocean DA and the archiving fails.  Could that also be added here?

DavidHuber-NOAA replied:
Yeah, I should be able to use that trigger instead.  I'll mimic the updates in #2560 and retest `AOWCDA`.

DavidHuber-NOAA replied:
Added and tested.

CatherineThomas-NOAA replied:
Looks good @DavidHuber-NOAA. Thanks!

DavidHuber-NOAA committed:
Use aero cdump variables to determine archiving

CoryMartin-NOAA commented on parm/config/gfs/config.fcst:
```diff
@@ -13,7 +13,7 @@ case ${WAVE_CDUMP} in
+# Turn off aerosols if not used for this CDUMP
 case ${AERO_FCST_CDUMP} in
   both | "${CDUMP/enkf}" ) ;; # Don't change
   *) DO_AERO="NO" ;; # Turn waves off
```
one other waves -> aerosols

DavidHuber-NOAA replied:
Thanks again.  Fixed.

CoryMartin-NOAA reviewed (commented):
I only reviewed the aero APP stuff, so I don't want to click "approve" but consider that section "approved"

DavidHuber-NOAA committed:
Replace waves with aerosols in config.fcst

JessicaMeixner-NOAA commented on parm/archive/gfswave.yaml.j2:
```diff
@@ -0,0 +1,19 @@
+            - "{{ COM_WAVE_GRID | relpath(ROTDIR) }}/{{ head }}*.*.f{{ '%03d' % fh }}.grib2"
+            - "{{ COM_WAVE_GRID | relpath(ROTDIR) }}/{{ head }}*.*.f{{ '%03d' % fh }}.grib2.idx"
+            {% endfor %}
+        optional:
```
Apologies, I haven't read through the full PR yet so this might be obvious from other parts of the code, but what is the difference between mandatory and optional?

DavidHuber-NOAA replied:
Mandatory files will raise a fatal error if they are not found during archiving.  Missing optional files will print warnings to the log file.

JessicaMeixner-NOAA replied:
As long as we have point output, I would think these should be mandatory too...

DavidHuber-NOAA replied:
Speaking to that, I think I "required" makes more sense than "mandatory", so I will update that.

JessicaMeixner-NOAA replied:
so {{ head }} in this case is just gfswave?  We do have a few other tar files that are only occationally created if DOIBP_WAV=YES, but we'd want those archived if they are created.

DavidHuber-NOAA replied:
Wave point data files have been made required.  The change was tested and works for a forecast-only S2SW case.

DavidHuber-NOAA replied:
Yes, that's what it is here.  If there's a case where additional data needs to be included in this file that does not start with `gfswave`, we can find a solution to add it.  `head` isn't a hard-fast prefix for all files in a given template.

DavidHuber-NOAA commented:
The issue description has been updated.

DavidHuber-NOAA committed:
Change mandatory to required

DavidHuber-NOAA committed:
Make wave point files required

JessicaMeixner-NOAA commented on parm/archive/gfswave.yaml.j2:
```diff
@@ -0,0 +1,18 @@
+            - "{{ COM_WAVE_GRID | relpath(ROTDIR) }}/{{ head }}*.*.f{{ '%03d' % fh }}.grib2"
+            - "{{ COM_WAVE_GRID | relpath(ROTDIR) }}/{{ head }}*.*.f{{ '%03d' % fh }}.grib2.idx"
+            {% endfor %}
+            - "{{ COM_WAVE_STATION | relpath(ROTDIR) }}/{{ head }}bull_tar"
```
If    DOIBP_WAV='YES' in https://github.com/NOAA-EMC/global-workflow/blob/develop/parm/config/gfs/config.wavepostsbs#L16 

Then we'll also need to archive: 
gfswave.t12z.ibpbull_tar  gfswave.t12z.ibpcbull_tar  gfswave.t12z.ibp_tar  

This isn't turned on in tests but we'll turn it back on eventually.  This is only something that's done for GFS not GEFS or GDAS etc.

DavidHuber-NOAA replied:
Alright, I added those product to `gfswave.yaml.j2` wrapped within a `{% if DOIBP_WAV %}`.  Rerunning the `gfsarch` job ran successfully without attempting to archive these files.

JessicaMeixner-NOAA replied:
Thanks @DavidHuber-NOAA ! This is nice and it's great to see an example of how to add a configuration If yes/no to one of these files as well.

aerorahul commented on ush/python/pygfs/task/archive.py:
```diff
@@ -0,0 +1,396 @@
+                    chgrp, get_gid, logit, mkdir_p, parse_j2yaml, rm_p, strftime,
+                    to_YMD, to_YMDH, Template, TemplateConstants)
+from yaml import CLoader as Loader
+from yaml import load
```
Are these used?

DavidHuber-NOAA replied:
No, I'll remove them.

DavidHuber-NOAA replied:
Removed.

aerorahul commented on ush/python/pygfs/task/archive.py:
```diff
@@ -0,0 +1,396 @@
+import glob
+import os
+import shutil
+from datetime import timedelta
```
Check if this is used.

DavidHuber-NOAA replied:
Nope, removed.

aerorahul commented on ush/python/pygfs/task/archive.py:
```diff
@@ -0,0 +1,396 @@
+from logging import getLogger
+from typing import Any, Dict, List
+
+import numpy as np
```
Check if used?

DavidHuber-NOAA replied:
They're not.  I'll remove them.

DavidHuber-NOAA committed:
Add wave IBP data

DavidHuber-NOAA committed:
Split archive execute method, determine commands (htar/tar, chmod, etc) in configure method

DavidHuber-NOAA committed:
Remove unused numpy, timedelta modules

DavidHuber-NOAA committed:
Split enkfgdas and enkfgfs master templates; create one tarball at a time.

aerorahul commented on parm/archive/chem.yaml.j2:
```diff
@@ -0,0 +1,7 @@
+    chem:
```
Is there a reason this (and others) are indented?
Rather than indenting the yaml here, we should use the `{% filter indent(width=4) %}` where this (and other) yamls are included.  That way, if the nesting changes in the parent, these yamls don't have to be updated.
E.g.
```jinja
{% filter indent(width=4) %}
{% include "chem.yaml.j2" %}
{% endfilter %}

DavidHuber-NOAA replied:
We discussed this in person and I made the changes in the last commit.  Thanks for the suggestion.

aerorahul commented on jobs/rocoto/arch.sh:
```diff
@@ -8,6 +8,12 @@ source "${HOMEgfs}/ush/preamble.sh"
+###############################################################
+# setup python path for workflow utilities and tasks
+wxflowPATH="${HOMEgfs}/ush/python:${HOMEgfs}/ush/python/wxflow/src"
+PYTHONPATH="${PYTHONPATH:+${PYTHONPATH}:}${wxflowPATH}"
```
```suggestion
PYTHONPATH="${PYTHONPATH:+${PYTHONPATH}:}${HOMEgfs}/ush/python"
```
I think this should work.

DavidHuber-NOAA replied:
That's slick, I'll give it a try.

DavidHuber-NOAA replied:
Updated and tested.

aerorahul commented on jobs/rocoto/earc.sh:
```diff
@@ -8,6 +8,12 @@ source "${HOMEgfs}/ush/preamble.sh"
+###############################################################
+# setup python path for workflow utilities and tasks
+wxflowPATH="${HOMEgfs}/ush/python:${HOMEgfs}/ush/python/wxflow/src"
+PYTHONPATH="${PYTHONPATH:+${PYTHONPATH}:}${wxflowPATH}"
```
```suggestion
PYTHONPATH="${PYTHONPATH:+${PYTHONPATH}:}${HOMEgfs}/ush/python"
```

DavidHuber-NOAA replied:
Updated and tested.

aerorahul commented on parm/archive/enkf.yaml.j2:
```diff
@@ -0,0 +1,55 @@
+    enkf:
+        {% set head = RUN + ".t" + cycle_HH + "z." %}
```
```suggestion
```
`head` should be defined in `master_enkf.yaml.j2` along with the other variables e.g. `cycle_YMDH`, etc.

DavidHuber-NOAA replied:
OK, I'll move them.

DavidHuber-NOAA replied:
Removed.

aerorahul commented on parm/archive/enkf_grp.yaml.j2:
```diff
@@ -0,0 +1,46 @@
+    enkf_grp:
+        {% set head = RUN + ".t" + cycle_HH + "z." %}
```
```suggestion
```
Similar to `enkf.yaml.j2`, this could be in `master_enkf.yaml.j2`

DavidHuber-NOAA replied:
Removed.

aerorahul commented on parm/archive/enkf_restarta_grp.yaml.j2:
```diff
@@ -0,0 +1,32 @@
+    enkf_restarta_grp:
+        {% set head = RUN + ".t" + cycle_HH + "z." %}
```
```suggestion
```
As above.

DavidHuber-NOAA replied:
Removed.

aerorahul commented on parm/archive/enkf_restartb_grp.yaml.j2:
```diff
@@ -0,0 +1,21 @@
+    enkf_restartb_grp:
+        {% set head = RUN + ".t" + cycle_HH + "z." %}
```
```suggestion
```
And here.

DavidHuber-NOAA replied:
Removed.

DavidHuber-NOAA committed:
Merge remote-tracking branch 'emc/develop' into feature/archive_yaml

DavidHuber-NOAA committed:
Handle indention at the master level

DavidHuber-NOAA commented:
@aerorahul Per our conversation yesterday, I updated the archive templates so tabbing is controlled at the master level.

DavidHuber-NOAA commented:
All changes have been tested in the `C48_S2SW`, `C96_atmaerosnowDA`, and `C48mx500_3DVarAOWCDA` test cases as well as an extended (25.5-cycle) `C96C48_hybatmDA` test case.

aerorahul commented on ush/python/pygfs/task/archive.py:
```diff
@@ -0,0 +1,383 @@
+
+        # Generate tarball
+        if len(atardir_set.fileset) == 0:
+            print(f"WARNING skipping would-be empty archive {atardir_set.target}.")
```
Use logger.
```suggestion
            logger.warning(f"WARNING: skipping would-be empty archive {atardir_set.target}.")
```

DavidHuber-NOAA replied:
Fixed.

aerorahul commented on ush/python/pygfs/task/archive.py:
```diff
@@ -0,0 +1,383 @@
+                for item in atardir_set.optional:
+                    glob_set = glob.glob(item)
+                    if len(glob_set) == 0:
+                        print(f"WARNING: optional file/glob {item} not found!")
```
Use logger

DavidHuber-NOAA replied:
Fixed.

aerorahul commented on ush/python/pygfs/task/archive.py:
```diff
@@ -0,0 +1,383 @@
+        file_list : List
+            List of files to add to an archive
+        """
+        import tarfile
```
Imports should be at the top

DavidHuber-NOAA replied:
Moved.

DavidHuber-NOAA committed:
Merge remote-tracking branch 'emc/develop' into feature/archive_yaml

DavidHuber-NOAA committed:
Consolidated enkf heads to master template

DavidHuber-NOAA committed:
Set PYTHONPATH to head of ush/python

DavidHuber-NOAA committed:
Clean up archive.py

DavidHuber-NOAA reviewed (commented):
@aerorahul Thanks for the suggestions.  I've updated and tested the branch.

CatherineThomas-NOAA reviewed (commented):
It looks like my previous comments were stuck in "pending" so I'm not sure if you could actually see them.  Hopefully you can now.

DavidHuber-NOAA committed:
Fix ocean archiving

DavidHuber-NOAA committed:
Removed unused COM, updated gdasocean template

DavidHuber-NOAA committed:
Explicitized gdasocean_analysis template

DavidHuber-NOAA committed:
Clean up some of the templates

DavidHuber-NOAA committed:
Merge remote-tracking branch 'emc/develop' into feature/archive_yaml

aerorahul reviewed (approved):
looks good.
thanks for working through the reviewer comments.

@DavidHuber-NOAA 
Would it be possible to run a few CI cases and pointing to the paths to `ARCDIR` and `ATARDIR` locally and on HPSS so that the reviewers (and others), can validate these datasets?

DavidHuber-NOAA commented:
@aerorahul Sure, I'll run the `C48_S2SW`, `C48mx500_3DVarAOWCDA`, `C96_atmaerosnowDA`, `C96C48_ufs_hybatmDA`, and `C96C48_hybatmDA` cases on Hera and extend the `C96C48_hybatmDA` case to 5 cycles with `ARCH_WARMICFREQ=1` so the GDAS restart archives are generated.

aerorahul commented:
> @aerorahul Sure, I'll run the `C48_S2SW`, `C48mx500_3DVarAOWCDA`, `C96_atmaerosnowDA`, `C96C48_ufs_hybatmDA`, and `C96C48_hybatmDA` cases on Hera and extend the `C96C48_hybatmDA` case to 5 cycles with `ARCH_WARMICFREQ=1` so the GDAS restart archives are generated.

Thanks @DavidHuber-NOAA 
When done, please post the paths so they can be shared with @CatherineThomas-NOAA @JessicaMeixner-NOAA and others

DavidHuber-NOAA committed:
Merge remote-tracking branch 'emc/develop' into feature/archive_yaml

DavidHuber-NOAA committed:
Do not archive ocean analysis on the first cycle

DavidHuber-NOAA committed:
Fix restart copies in forecast_postdet.sh

DavidHuber-NOAA committed:
Add archive support for jedi ATM; fix reset time offsets when DOIAU=NO

DavidHuber-NOAA committed:
Reverted restart offset for enkf (does not change with IAU)

JessicaMeixner-NOAA commented:
Confirmed that that the new archive job does catch the error that was reported in https://github.com/NOAA-EMC/global-workflow/issues/2242

WalterKolczynski-NOAA commented on parm/config/gefs/config.base:
```diff
@@ -81,6 +81,10 @@ export CHGRP_CMD="@CHGRP_CMD@"
 export NCDUMP="${NETCDF:-${netcdf_c_ROOT:-}}/bin/ncdump"
 export NCLEN="${HOMEgfs}/ush/getncdimlen"
 
+# Stop gap fix for wgrib with spack-stack 1.6.0
+# TODO Remove this when spack-stack issue #1097 is resolved
+export WGRIB="${WGRIB:-${grib_util_ROOT:-}/wgrib}"
```
The module files are already defining `WGRIB`, why is it needed here?

DavidHuber-NOAA replied:
The grib-util module file is defining `WGRIB2` by accident.  See line 45 of `/scratch1/NCEPDEV/nems/role.epic/spack-stack/spack-stack-1.6.0/envs/gsi-addon-dev-rocky8/install/modulefiles/intel/2021.5.0/grib-util/1.3.0.lua`:
```lua
setenv("WGRIB2", "/scratch1/NCEPDEV/nems/role.epic/spack-stack/spack-stack-1.6.0/envs/gsi-addon-dev-rocky8/install/intel/2021.5.0/grib-util-1.3.0-ms2iw3h/bin/wgrib2")
```
The spack-stack developers are working on fixing this in the code and then will update the module files locally.  Until then, this is needed for a couple of the wave scripts.

WalterKolczynski-NOAA replied:
_Our_ modulefiles are defining WGRIB2 because the spack-stack ones weren't. See https://github.com/NOAA-EMC/global-workflow/blob/develop/modulefiles/module_base.hera.lua#L43 or any of the other spack-based module_base files.

JessicaMeixner-NOAA replied:
Should this be in ush/load_ufswm_modules.sh instead? We already have some updates there for WGRIB2

DavidHuber-NOAA replied:
@WalterKolczynski-NOAA Agreed, our module file is doing that for `WGRIB2`, but not `WGRIB`.  I think moving this line over to the `module_base`s would be fine.

WalterKolczynski-NOAA replied:
OH! Sorry, I was misreading it. Yeah, it should be in the module file.

DavidHuber-NOAA replied:
Moved.

WalterKolczynski-NOAA commented on parm/config/gefs/config.fcst:
```diff
@@ -14,6 +14,12 @@ case ${WAVE_CDUMP} in
 
+# Turn off waves if not used for this CDUMP
+case ${AERO_FCST_CDUMP} in
+  both | "${CDUMP/enkf}" ) ;; # Don't change
```
```suggestion
  both | "${RUN/enkf}" ) ;; # Don't change
```

WalterKolczynski-NOAA commented on parm/config/gefs/config.fcst:
```diff
@@ -14,6 +14,12 @@ case ${WAVE_CDUMP} in
   *) DO_WAVE="NO" ;; # Turn waves off
 esac
 
+# Turn off waves if not used for this CDUMP
```
```suggestion
# Turn off waves if not used for this RUN
```

DavidHuber-NOAA replied:
Why do aerosols need to key off of RUN and waves off of CDUMP?  Or should they both key off of RUN?

WalterKolczynski-NOAA replied:
They should both key off `RUN`. We are trying to eliminate CDUMP except in a select few instances related to observations (see #1299).

DavidHuber-NOAA replied:
Changed to RUN.

WalterKolczynski-NOAA commented on parm/config/gfs/config.base:
```diff
@@ -110,6 +110,10 @@ export CHGRP_CMD="@CHGRP_CMD@"
 export NCDUMP="${NETCDF:-${netcdf_c_ROOT:-}}/bin/ncdump"
 export NCLEN="${HOMEgfs}/ush/getncdimlen"
 
+# Stop gap fix for wgrib with spack-stack 1.6.0
+# TODO Remove this when spack-stack issue #1097 is resolved
+export WGRIB="${WGRIB:-${grib_util_ROOT:-}/wgrib}"
```
Same as before; this is already in the module files.

DavidHuber-NOAA replied:
Moved.

WalterKolczynski-NOAA commented on parm/config/gfs/config.base:
```diff
@@ -181,8 +182,11 @@ export DO_WAVE="NO"
 export DO_OCN="NO"
 export DO_ICE="NO"
 export DO_AERO="NO"
+export AERO_FCST_CDUMP="" # When to run aerosol forecast: gdas, gfs, or both
+export AERO_ANL_CDUMP="" # When to run aerosol analysis: gdas, gfs, or both
```
Works for now to get it functional, but we should change the python so the previous method works correctly (and get rid of `WAVE_CDUMP` too).

DavidHuber-NOAA replied:
Noted.

DavidHuber-NOAA committed:
Add missing variables

DavidHuber-NOAA committed:
Fix VFYARC archiving

DavidHuber-NOAA committed:
Fix offsets for ENKF IAU

WalterKolczynski-NOAA commented on parm/config/gfs/config.earc:
```diff
@@ -8,7 +8,22 @@ echo "BEGIN: config.earc"
 . $EXPDIR/config.resources earc
 
-export NMEM_EARCGRP=10
+# Set the number of ensemble members to archive per earc job
+case "${CASE_ENS}" in
+   "C48" | "C96")
+      export NMEM_EARCGRP=80
+      ;;
+   "C192")
+      export NMEM_EARCGRP=20
+      ;;
+   "C384" | "C768")
+      export NMEM_EARCGRP=10
+      ;;
+   *)
+      echo "FATAL ERROR: Unknown ensemble resolution ${CASE_ENS}, ABORT!"
+      exit 1
+      ;;
+esac
```
We need a C1152 case. That is already being run for HR3.

DavidHuber-NOAA replied:
Added an 1152 case, though I haven't tested it.  4 is a fairly conservative estimate.

WalterKolczynski-NOAA commented on parm/config/gfs/config.fcst:
```diff
@@ -14,6 +14,12 @@ case ${WAVE_CDUMP} in
   *) DO_WAVE="NO" ;; # Turn waves off
 esac
 
+# Turn off aerosols if not used for this CDUMP
```
```suggestion
# Turn off aerosols if not used for this RUN
```

DavidHuber-NOAA replied:
Updated

WalterKolczynski-NOAA commented on parm/config/gfs/config.fcst:
```diff
@@ -14,6 +14,12 @@ case ${WAVE_CDUMP} in
 
+# Turn off aerosols if not used for this CDUMP
+case ${AERO_FCST_CDUMP} in
+  both | "${CDUMP/enkf}" ) ;; # Don't change
```
```suggestion
  both | "${RUN/enkf}" ) ;; # Don't change
```

DavidHuber-NOAA replied:
Updated.

DavidHuber-NOAA committed:
Fix ocean/ice GDAS products

WalterKolczynski-NOAA commented on parm/config/gfs/config.wave:
```diff
@@ -91,16 +91,10 @@ export WAVEWND_DID=
-  export FHMAX_WAV=${FHMAX:-9}
-else
-  export FHMAX_WAV=${FHMAX_GFS}
+if [[ "${CDUMP}" = "gfs" ]]; then
```
```suggestion
if [[ "${RUN}" == "gfs" ]]; then
```

DavidHuber-NOAA replied:
Updated.

DavidHuber-NOAA committed:
Wrap optional files in appropriate flags

DavidHuber-NOAA commented:
@aerorahul @JessicaMeixner-NOAA @CatherineThomas-NOAA @jiaruidong2017

I finished running 5 CI test cases on Hera.  Could you please check the contents of the tarballs from these tests to verify that all necessary files are archived?  The tarballs have been pushed to HPSS under the following subdirectories in `/NCEPDEV/emc-global/1year/David.Huber/HERA/scratch`:

```console
arch_c48_s2sw
arch_c48mx500_3dvaraowcda
arch_c96_atmaerosnowda
arch_c96c48_hybatmda
arch_c96c48_ufs_hybatmda
```

For convenience, I have downloaded, listed, and extracted all of the tarballs onto Hera under the same subdirectories in `/scratch2/NCEPDEV/stmp1/David.Huber/archive_rotdir`.

The COMROOT for each of these experiments can be found in `/scratch1/NCEPDEV/global/David.Huber/para/comrot`.  Please let me know if you need anything else.  Thanks!

DavidHuber-NOAA committed:
Use RUN instead of CDUMP for aerosols.

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

DavidHuber-NOAA committed:
Added an 1152 case for earc

DavidHuber-NOAA committed:
Replace CDUMP with RUN for wave, aerosol configs

DavidHuber-NOAA committed:
Move WGRIB def to module_base

DavidHuber-NOAA committed:
Added missing ;;.

JessicaMeixner-NOAA commented:
@DavidHuber-NOAA thanks for adding the paths to the hpss locations.  Admittedly, I don't remember the contents of all the existing tar-files.  Do you know if there's an HPSS of a recent PR run that we could grab to look at the tar-files for?  I'm doing checks of what I can remember to look for.

WalterKolczynski-NOAA reviewed (approved):
Conditionally approved pending successful completion of manual & CI tests.

WalterKolczynski-NOAA commented:
@DavidHuber-NOAA Let me know when your testing is successful, then I will start CI.

DavidHuber-NOAA commented:
@JessicaMeixner-NOAA Yes, the CI tests for #2588 should work well for that purpose.  You can find the archives in `/NCEPDEV/emc-global/1year/Terry.McGuinness/HERA/scratch` under the following subdirectories:
```console
C48mx500_3DVarAOWCDA_ef56339a
C96C48_hybatmDA_ef56339a
C96_atmaerosnowDA_ef56339a
C48_S2SW_ef56339a
```
Note that the `C96C48_ufs_hybatmDA` test is not currently run as part of the CI suite.  I can run this separately using the current `hpssarch_gen.sh` method if you like.

Also, I reworked how often restarts are generated.  Previously, these were generated every cycle, though I think that was not the intent.  Instead, they should be generated (according to config.base) every `ARCH_WARMICFREQ` (for warm start gdas ICs) and every `ARCH_FCSTICFREQ` (for forecast-only gdas/gfs ICs) days on the `ARCH_CYC` cycle.  By default, warm ICs are now generated every 4 days and forecast only ICs every day, both at 00z.  However, for these tests, the warm start ICs are generated after one day (cycle `2021122200` for `arch_c96c48_hybatmda` and `2024022500` for `arch_c96c48_ufs_hybatmda`).

CatherineThomas-NOAA commented on parm/archive/gdas_restarta.yaml.j2:
```diff
@@ -0,0 +1,49 @@
+    required:
+        {% if current_cycle != SDATE and MODE == "cycled" %}
+        - "{{ COM_ATMOS_ANALYSIS | relpath(ROTDIR) }}/{{ head }}atminc.nc"
+        {% for iauhr in IAUHRS if iauhr != 6 %}
```
These should probably be IAUFHRS, not IAUHRS.

DavidHuber-NOAA replied:
Thanks for this catch.  I'll fix this and regen the archives.

CatherineThomas-NOAA commented on parm/archive/gfs_netcdfa.yaml.j2:
```diff
@@ -0,0 +1,16 @@
+        - "{{ COM_ATMOS_ANALYSIS | relpath(ROTDIR) }}/{{ head }}atmanl.nc"
+        - "{{ COM_ATMOS_ANALYSIS | relpath(ROTDIR) }}/{{ head }}sfcanl.nc"
+        - "{{ COM_ATMOS_ANALYSIS | relpath(ROTDIR) }}/{{ head }}atminc.nc"
+        {% for iauhr in IAUHRS if iauhr != 6 %}
```
Same comment as gdas_restarta --> IAUFHRS

CatherineThomas-NOAA commented on parm/archive/gdas.yaml.j2:
```diff
@@ -0,0 +1,132 @@
+        {% endif %}
+        {% endif %}
+        {% for fhr in range(FHMIN, FHMAX + FHOUT, FHOUT) %}
+        - "{{ COM_ATMOS_MASTER | relpath(ROTDIR) }}/{{ head }}sfluxgrbf{{ '%03d' % fhr }}.grib2"
```
The sflux files are saved earlier in this yaml (Lines 95/96)

DavidHuber-NOAA replied:
Yep, I'll delete these from optional.

CatherineThomas-NOAA commented on parm/archive/gdas.yaml.j2:
```diff
@@ -0,0 +1,132 @@
+    optional:
+        {% if current_cycle != SDATE and MODE == "cycled" %}
+        {% if not DO_JEDIATMVAR %}
+        - "{{ COM_ATMOS_ANALYSIS | relpath(ROTDIR) }}/{{ head }}gsistat"
```
The gsistat file should be required if cycling (still only for DO_JEDIATMVAR=NO).

CatherineThomas-NOAA commented on parm/archive/gfsa.yaml.j2:
```diff
@@ -0,0 +1,62 @@
+        - "{{ COM_ATMOS_ANALYSIS | relpath(ROTDIR) }}/{{ head }}atmvar.yaml"
+        - "{{ COM_ATMOS_ANALYSIS | relpath(ROTDIR) }}/{{ head }}atmstat"
+        {% else %}
+        - "{{ COM_ATMOS_ANALYSIS | relpath(ROTDIR) }}/{{ head }}gsistat"
```
Same comment about the gsistat being required.  

@RussTreadon-NOAA: Should the atmvar.yaml and atmstat files similarly be required for DO_JEDIATMVAR=YES?

CatherineThomas-NOAA commented:
General discussion point:  For the gdas tarballs (enkfgdas_grpN.tar, enkfgdas.tar, and gdas.tar), history/grib files up to F009 are being saved every hour instead of every 3 hours as previous.  Was this intentional?

JessicaMeixner-NOAA commented on parm/archive/gdasocean.yaml.j2:
```diff
@@ -0,0 +1,9 @@
+    target: "{{ ATARDIR }}/{{ cycle_YMDH }}/gdasocean.tar"
+    required:
+        {% for fhr in range(FHMIN, FHMAX + 1, FHMAX) %}
+        - "{{ COM_OCEAN_HISTORY | relpath(ROTDIR) }}/{{ head }}inst.f{{ '%03d' % fhr }}.nc"
```
@AndrewEichmann-NOAA which forecast hours are supposed to be archived here?  

Currently we are getting: 
gdas.20210324/12/model_data/ocean/history/gdas.ocean.t12z.inst.f000.nc
gdas.20210324/12/model_data/ocean/history/gdas.ocean.t12z.inst.f009.nc

and not the f003 and f006, is that okay or should those forecast hours also be archived?  I believe with your fix in https://github.com/NOAA-EMC/global-workflow/pull/2594 f003 and f006 would also be archived.

DavidHuber-NOAA replied:
Yeah, I think that should be `FHOUT_OCNICE`, not `FHMAX`, for the step size in the for loop.  I'll update and test.

JessicaMeixner-NOAA commented on parm/archive/gdasice.yaml.j2:
```diff
@@ -0,0 +1,9 @@
+    required:
+        {% for fhr in range(FHOUT_OCNICE, FHMAX+1, FHOUT_OCNICE) %}
+        - "{{ COM_ICE_HISTORY | relpath(ROTDIR) }}/{{ head }}inst.f{{ '%03d' % fhr }}.nc"
+        {% endfor %}
```
@DavidHuber-NOAA in the develop version, we have one more file in this tar-file  gdas.20210324/12/model_data/ice/history/gdas.ice.t12z.ic.nc  that is missing in /scratch2/NCEPDEV/stmp1/David.Huber/archive_rotdir/arch_c48mx500_3dvaraowcda/2021032412/gdasice.tar.listing

From: /scratch1/NCEPDEV/global/CI/2588/RUNTESTS/COMROOT/C48mx500_3DVarAOWCDA_ef56339a/logs/2021032412/gdasarch.log

```
++ exglobal_archive.sh[289]: cat /scratch1/NCEPDEV/stmp2/Terry.McGuinness/RUNDIRS/C48mx500_3DVarAOWCDA_ef56339a/arch.576624/gdasice.txt
+ exglobal_archive.sh[289]: htar -P -cvf /NCEPDEV/emc-global/1year/Terry.McGuinness/HERA/scratch/C48mx500_3DVarAOWCDA_ef56339a/2021032412/gdasice.tar gdas.20210324/12//model_data/ice/history/gdas.ice.t12z.ic.nc gdas.20210324/12//model_data/ice/history/gdas.ice.t12z.inst.f003.nc gdas.20210324/12//model_data/ice/history/gdas.ice.t12z.inst.f006.nc gdas.20210324/12//model_data/ice/history/gdas.ice.t12z.inst.f009.nc
[connecting to hpsscore1.fairmont.rdhpcs.noaa.gov/1217]
HTAR: a   gdas.20210324/12/model_data/ice/history/gdas.ice.t12z.ic.nc
HTAR: a   gdas.20210324/12/model_data/ice/history/gdas.ice.t12z.inst.f003.nc
HTAR: a   gdas.20210324/12/model_data/ice/history/gdas.ice.t12z.inst.f006.nc
HTAR: a   gdas.20210324/12/model_data/ice/history/gdas.ice.t12z.inst.f009.nc




```

DavidHuber-NOAA replied:
Thanks!  I'll add in the ic file as well.

JessicaMeixner-NOAA commented on parm/archive/gdasocean_analysis.yaml.j2:
```diff
@@ -0,0 +1,24 @@
+    {% set head = "gdas.t" + cycle_HH + "z." %}
+    name: "GDASOCEAN_ANALYSIS"
+    target: "{{ ATARDIR }}/{{ cycle_YMDH }}/gdasocean_analysis.tar"
+    required:
```
I see three missing files when I compare 
/scratch1/NCEPDEV/global/CI/2588/RUNTESTS/COMROOT/C48mx500_3DVarAOWCDA_ef56339a/logs/2021032418/gdasarch.log
with 
/scratch2/NCEPDEV/stmp1/David.Huber/archive_rotdir/arch_c48mx500_3dvaraowcda/2021032418/gdasocean_analysis.tar.listing
They are: 

HTAR: a   gdas.20210324/18/analysis/ocean/gdas.t18z.ocn.adt_rads_all.stats.csv
HTAR: a   gdas.20210324/18/analysis/ocean/gdas.t18z.ocn.icec_amsr2_north.stats.csv
HTAR: a   gdas.20210324/18/analysis/ocean/gdas.t18z.ocn.icec_amsr2_south.stats.csv

@AndrewEichmann-NOAA @guillaumevernieres can you confirm that these should be archived.

guillaumevernieres replied:
Yes, they should.

DavidHuber-NOAA replied:
Alright, thanks, I'll add them in.

JessicaMeixner-NOAA commented:
@DavidHuber-NOAA I've taken a look at both C48mx500_3DVarAOWCDA_ef56339a and C48_S2SW_ef56339a

The C48_S2SW is as expected as far as I can tell - it is missing the restart tar-file as you have explained.   The  C48mx500_3DVarAOWCDA, I have a few questions about missing files which I made specific comments in the yaml files and tagged the marine DA pocs for as well.   We're also missing all of the restart tars for this test, but I'm assuming that's again intentional.

DavidHuber-NOAA commented:
> General discussion point: For the gdas tarballs (enkfgdas_grpN.tar, enkfgdas.tar, and gdas.tar), history/grib files up to F009 are being saved every hour instead of every 3 hours as previous. Was this intentional?

@CatherineThomas-NOAA I did do this intentionally, but now that I have a fresh think about it, it will add a significant amount of data to the ensemble tarballs.  If there isn't a use case, I am happy to revert it to saving every 3 hours (or perhaps keying on `IAUFHRS`).

DavidHuber-NOAA committed:
Fix IAUFHRS

DavidHuber-NOAA committed:
Make gsistat mandatory; remove duplicate flux files

CatherineThomas-NOAA commented:
> I did do this intentionally, but now that I have a fresh think about it, it will add a significant amount of data to the ensemble tarballs. If there isn't a use case, I am happy to revert it to saving every 3 hours (or perhaps keying on IAUFHRS).

Thanks @DavidHuber-NOAA. If it's alright with others, I think it's better to keep it at 3 hourly for now and if there comes a use case, we can add it in later.  @JessicaMeixner-NOAA and I plan to do a thorough accounting of what's getting saved a little further down the line, so we can revisit it then while keeping the tarball sizes down for now.

DavidHuber-NOAA committed:
Fix ocean time step to archive

DavidHuber-NOAA committed:
Add missing ice IC file to archive.

DavidHuber-NOAA committed:
Add missing ocean CSVs.

DavidHuber-NOAA commented:
@CatherineThomas-NOAA Sounds good, I'll go with 3-hourly, then.

DavidHuber-NOAA committed:
Archive *gdas hourly outputs at 3-hour intervals

DavidHuber-NOAA commented:
@CatherineThomas-NOAA @JessicaMeixner-NOAA @guillaumevernieres I have addressed all of the comments and will start regenerating the tarballs.  Note that I will not be able to regenerate the first two cycles of the `*hybatmDA` tests as that data has been scrubbed by the workflow.

DavidHuber-NOAA committed:
Add missing keys and fix bugs

DavidHuber-NOAA committed:
Merge remote-tracking branch 'emc/develop' into feature/archive_yaml

DavidHuber-NOAA committed:
Flesh out enkf restarts.

DavidHuber-NOAA committed:
Key enkf atmf netcdf files off of IAUFHRS_ENKF

DavidHuber-NOAA committed:
Add missing logs and inc file to enkf*.tar

DavidHuber-NOAA committed:
Make atma files optional, correct log name

DavidHuber-NOAA committed:
Add sfcanl data files to gdas restart tarballs

DavidHuber-NOAA committed:
Move misplaced endif

DavidHuber-NOAA committed:
Add iaufhrs_enkf definition

DavidHuber-NOAA committed:
Merge remote-tracking branch 'emc/develop' into feature/archive_yaml

DavidHuber-NOAA commented:
@WalterKolczynski-NOAA This PR is now ready for CI testing.

DavidHuber-NOAA commented:
@WalterKolczynski-NOAA @TerrenceMcGuinness-NOAA It looks like the ICs failed to stage correctly for the AOWCDA test case on on WCOSS2.  I'll take a deeper look as to what happened.

WalterKolczynski-NOAA commented:
> @WalterKolczynski-NOAA @TerrenceMcGuinness-NOAA It looks like the ICs failed to stage correctly for the AOWCDA test case on on WCOSS2. I'll take a deeper look as to what happened.

That case shouldn't be running on WCOSS. @TerrenceMcGuinness-NOAA recently made a change to the case-skipping stuff, so maybe that is malfunctioning.

TerrenceMcGuinness-NOAA commented:
Oh yes!  We need to update the CI scripts dir on WCOSS2 with the updates for host skipping.

WalterKolczynski-NOAA commented:
Hera is some issue running `rocotostat`, though I think the underlying problem is with the disk (maybe stmp is full?)

DavidHuber-NOAA commented:
@TerrenceMcGuinness-NOAA rocotostat.py is failing on Hera:
```console
+ /scratch1/NCEPDEV/global/CI/2491/gfs/ci/scripts/run-check_ci.sh /scratch1/NCEPDEV/global/CI/2491 C48_ATM_629bf6fd gfs
Source modules.
Running "module reset". Resetting modules to system default. The following $MODULEPATH directories have been removed: /scratch[1](https://jenkins.epic.oarcloud.noaa.gov/job/global-workflow/job/EMC-Global-Pipeline/job/PR-2491/2/pipeline-console/?start-byte=0&selected-node=322#log-1)/NCEPDEV/global/Terry.McGuinness/global-workflow/modulefiles
cd /scratch1/NCEPDEV/global/CI/[2](https://jenkins.epic.oarcloud.noaa.gov/job/global-workflow/job/EMC-Global-Pipeline/job/PR-2491/2/pipeline-console/?start-byte=0&selected-node=322#log-2)491/RUNTESTS/EXPDIR/C48_ATM_629bf6fd
Launch experiment with Rocoto.
Run rocotorun.
Gather Rocoto statistics
(C48_ATM_629bf6fd on Hera)
	Total Cycles: 1
	Number Cycles done: 0
	State: RUNNING
Run rocotorun.
Gather Rocoto statistics
(C48_ATM_629bf6fd on Hera)
	Total Cycles: 1
	Number Cycles done: 0
	State: RUNNING
Run rocotorun.
Gather Rocoto statistics
(C48_ATM_629bf6fd on Hera)
	Total Cycles: 1
	Number Cycles done: 0
	State: RUNNING
Run rocotorun.
Gather Rocoto statistics
(C48_ATM_629bf6fd on Hera)
	Total Cycles: 1
	Number Cycles done: 0
	State: RUNNING
Run rocotorun.
Gather Rocoto statistics
(C48_ATM_629bf6fd on Hera)
	Total Cycles: 1
	Number Cycles done: 0
	State: RUNNING
Run rocotorun.
Gather Rocoto statistics
05/[16](https://jenkins.epic.oarcloud.noaa.gov/job/global-workflow/job/EMC-Global-Pipeline/job/PR-2491/2/pipeline-console/?start-byte=0&selected-node=322#log-16)/24 20:18:26 UTC :: C48_ATM_629bf6fd.xml :: Never received the URI and/or pid of rocotodbserver.
05/16/24 20:[18](https://jenkins.epic.oarcloud.noaa.gov/job/global-workflow/job/EMC-Global-Pipeline/job/PR-2491/2/pipeline-console/?start-byte=0&selected-node=322#log-18):29 UTC :: C48_ATM_629bf6fd.xml :: ERROR! Could not launch database server process.
Traceback (most recent call last):
  File "/scratch1/NCEPDEV/global/CI/2491/gfs/ci/scripts/utils/rocotostat.py", line 85, in <module>
    rocoto_status = rocoto_statcount()
                    ^^^^^^^^^^^^^^^^^^
  File "/scratch1/NCEPDEV/global/CI/2491/gfs/ci/scripts/utils/rocotostat.py", line 59, in rocoto_statcount
    rocotostat_output = rocotostat(output=str)
                        ^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch1/NCEPDEV/global/CI/2491/gfs/ci/scripts/utils/wxflow/executable.py", line 230, in __call__
    raise ProcessError(f"Command exited with status {proc.returncode}:", long_msg)
wxflow.executable.ProcessError: Command exited with status 1:
'/apps/rocoto/1.3.7/bin/rocotostat' '-w' '/scratch1/NCEPDEV/global/CI/2491/RUNTESTS/EXPDIR/C48_ATM_629bf6fd/C48_ATM_629bf6fd.xml' '-d' '/scratch1/NCEPDEV/global/CI/2491/RUNTESTS/EXPDIR/C48_ATM_629bf6fd/C48_ATM_629bf6fd.db' '-s'
(C[48](https://jenkins.epic.oarcloud.noaa.gov/job/global-workflow/job/EMC-Global-Pipeline/job/PR-2491/2/pipeline-console/?start-byte=0&selected-node=322#log-48)_ATM_629bf6fd on Hera)
	Total Cycles: 
	Number Cycles done: 
	State: 
Experiment C48_ATM_629bf6fd Terminated with  tasks failed and  dead at Thu May 16 20:18:32 UTC 2024
Experiment C48_ATM_629bf6fd Terminated: **
script returned exit code 1
```

But the status on Hera was actually fine:
```
[David.Huber@hfe02 ~]$ rocotostat -w /scratch1/NCEPDEV/global/CI/2491/RUNTESTS/EXPDIR/C48_ATM_629bf6fd/C48_ATM_629bf6fd.xml -d /scratch1/NCEPDEV/global/CI/2491/RUNTESTS/EXPDIR/C48_ATM_629bf6fd/C48_ATM_629bf6fd.db
       CYCLE                    TASK                       JOBID               STATE         EXIT STATUS     TRIES      DURATION
================================================================================================================================
202103231200             gfsstage_ic                    60313053           SUCCEEDED                   0         1          24.0
202103231200                 gfsfcst                    60313237             RUNNING                   -         0           0.0
202103231200    gfsatmos_prod_f000-f006    druby://10.184.3.61:45189          SUBMITTING                   -         0           0.0
202103231200    gfsatmos_prod_f009-f015    druby://10.184.3.61:45189          SUBMITTING                   -         0           0.0
202103231200    gfsatmos_prod_f018-f024                           -                   -                   -         -             -
202103231200    gfsatmos_prod_f027-f033                           -                   -                   -         -             -
202103231200    gfsatmos_prod_f036-f042                           -                   -                   -         -             -
202103231200    gfsatmos_prod_f045-f051                           -                   -                   -         -             -
202103231200    gfsatmos_prod_f054-f060                           -                   -                   -         -             -
202103231200    gfsatmos_prod_f063-f069                           -                   -                   -         -             -
202103231200    gfsatmos_prod_f072-f078                           -                   -                   -         -             -
202103231200    gfsatmos_prod_f081-f087                           -                   -                   -         -             -
202103231200    gfsatmos_prod_f090-f096                           -                   -                   -         -             -
202103231200    gfsatmos_prod_f099-f105                           -                   -                   -         -             -
202103231200    gfsatmos_prod_f108-f114                           -                   -                   -         -             -
202103231200    gfsatmos_prod_f117-f120                           -                   -                   -         -             -
202103231200              gfstracker                           -                   -                   -         -             -
202103231200              gfsgenesis                           -                   -                   -         -             -
202103231200                 gfsarch                           -                   -                   -         -             -
202103231200              gfscleanup                           -                   -                   -         -             -
```

DavidHuber-NOAA commented:
Also, all of the stmp allocations are all under quota, as is global:
```
        Project: global
                Directory: /scratch1/NCEPDEV/global DiskInUse=337362 GB, Quota=387000 GB, Files=43340129, FileQUota=77400000
        Project: stmp
                Directory: /scratch1/NCEPDEV/stmp DiskInUse=684928 GB, Quota=700000 GB, Files=68787950, FileQUota=140000000
                Directory: /scratch2/NCEPDEV/stmp DiskInUse=531513 GB, Quota=700000 GB, Files=31714028, FileQUota=140000000
                Directory: /scratch2/NCEPDEV/stmp1 DiskInUse=531513 GB, Quota=700000 GB, Files=31714028, FileQUota=140000000
                Directory: /scratch1/NCEPDEV/stmp2 DiskInUse=684928 GB, Quota=700000 GB, Files=68787947, FileQUota=140000000
                Directory: /scratch2/NCEPDEV/stmp3 DiskInUse=531513 GB, Quota=700000 GB, Files=31714028, FileQUota=140000000
                Directory: /scratch1/NCEPDEV/stmp4 DiskInUse=684928 GB, Quota=700000 GB, Files=68787975, FileQUota=140000000
```

DavidHuber-NOAA commented:
Orion failed due to an srun error:

```
wxflow.executable.ProcessError: Command exited with status 99:
'srun' '-l' '--export=ALL' '-n' '1' '--cpus-per-task=1' '/work/noaa/stmp/mterry/RUNDIRS/C48_S2SWA_gefs_629bf6fd/oceanice_products.453382/ocnicepost.x'
```
I'll report this one to RDHPCS.

DavidHuber-NOAA committed:
Updating nc check script to verify time dimensions

DavidHuber-NOAA committed:
Ignore shellcheck warning

DavidHuber-NOAA committed:
Fixed ncfile variable name

DavidHuber-NOAA commented:
It looks like things are failing on Orion/Hercules due to a full stmp.

Meanwhile, the `earc` jobs are failing on WCOSS2 when trying to decorate the _create_fileset static method in archive.py.  This may be an issue of the older Python version on WCOSS2.  Will test and fix offline.

DavidHuber-NOAA committed:
Do not try to collect ecen, esfc logs on first cycle

DavidHuber-NOAA committed:
Move staticmethod for compatability with older Python versions

DavidHuber-NOAA commented:
@aerorahul @WalterKolczynski-NOAA @TerrenceMcGuinness-NOAA I tested the `C96C48_hybatmDA` case offline with `HPSSARCH=YES` on WCOSS2 successfully after a couple of modifications.  However, Hercules' and Orion's /work/noaa/stmp is still 5TB over quota.  Hera and WCOSS2 should be good to go, though.  Should I kick this off on those machines then come back to Hercules and Orion when stmp clears out?

WalterKolczynski-NOAA commented:
stmp still full on MSU

DavidHuber-NOAA commented:
`rocotostat` failed again on Hera.  It looks like rocoto had trouble reaching the DB server and is probably a transient issue:
```
05/17/24 17:53:26 UTC :: C48_S2SW_ffeaf759.xml :: Never received the URI and/or pid of rocotodbserver.
05/17/24 17:53:27 UTC :: C48_S2SW_ffeaf759.xml :: ERROR! Could not launch database server process.
```

Is it possible to force Jenkins to start up again without restarting the whole CI?

WalterKolczynski-NOAA commented:
> `rocotostat` failed again on Hera. It looks like rocoto had trouble reaching the DB server and is probably a transient issue:
> 
> ```
> 05/17/24 17:53:26 UTC :: C48_S2SW_ffeaf759.xml :: Never received the URI and/or pid of rocotodbserver.
> 05/17/24 17:53:27 UTC :: C48_S2SW_ffeaf759.xml :: ERROR! Could not launch database server process.
> ```
> 
> Is it possible to force Jenkins to start up again without restarting the whole CI?

Not currently, unfortunately. We should be able to harden the system so that a single `rocotostatus` fail doesn't sink things though.

I do wonder why we've been having problems suddenly over the last day or so. Might need to check with @christopherwharrop-noaa and/or RDHPCS admins.

WalterKolczynski-NOAA commented:
stmp usage has come down a _little_ bit and the Hera issue is hopefully ephemeral, so I'm going to take another crack at running the tests.

DavidHuber-NOAA commented:
Orion failed to load modules:
```
Begin oceanice_products.sh at Sat May 18 03:16:14 UTC 2024
+ oceanice_products.sh[11]: . /work2/noaa/stmp/CI/ORION/2491/gefs/ush/load_fv3gfs_modules.sh
++ load_fv3gfs_modules.sh[4]: [[ NO == \N\O ]]
++ load_fv3gfs_modules.sh[5]: echo 'Loading modules quietly...'
Loading modules quietly...
++ load_fv3gfs_modules.sh[6]: set +x
Lmod has detected the following error: Cannot load module
"stack-intel/2022.0.2" without these module(s) loaded:
   intel/2022.1.2

While processing the following module(s):
    Module fullname       Module Filename
    ---------------       ---------------
    stack-intel/2022.0.2  /work/noaa/epic/role-epic/spack-stack/orion/spack-stack-1.6.0/envs/gsi-addon-env/install/modulefiles/Core/stack-intel/2022.0.2.lua
    module_base.orion     /work2/noaa/stmp/CI/ORION/2491/gefs/modulefiles/module_base.orion.lua

End oceanice_products.sh at 03:16:14 with error code 1 (time elapsed: 00:00:00)
```
I saw this on Friday and reported it, but the admins saw no issues.  I'll see if I can recreate it and then rereport it.

DavidHuber-NOAA commented:
The issue on Orion appears to be resolved now, so I am rebooting CI on that machine.

On Hercules, processing stalled because some of the gfsfcst outputs appear to not have been generated, causing some of the early post jobs not to launch.  This may have been a disk space issue, but I'm not sure why the GFS wouldn't have failed in that case.  I'll reboot CI on that machine as well.

DavidHuber-NOAA committed:
Merge remote-tracking branch 'emc/develop' into feature/archive_yaml

DavidHuber-NOAA commented:
Restarting on Orion now.

DavidHuber-NOAA committed:
Separate stmp/ptmp for orion and hercules.

DavidHuber-NOAA committed:
Revert "Separate stmp/ptmp for orion and hercules."

This reverts commit 5aeccac58aeb6bb25fee37b8b73220502fac48e9.

WalterKolczynski-NOAA commented:
Issue with Orion is unrelated to this PR, and there is no reason to think this would work on Hercules but not Orion, so I am just going to merge this.

#2614 should resolve the CI issue.

aerorahul commented:
@CatherineThomas-NOAA @JessicaMeixner-NOAA @CoryMartin-NOAA @guillaumevernieres 
Could you please approve this PR?
